Mathlib/Algebra/GradedHMul.lean

```diff
@@ -0,0 +1,121 @@
+
+Part of the reason for this design choice is that if `M` is an additive monoid,
+`x : X a`, `y : Y 0`, we would usually want the product of `x` and `y` to be in
+`Z a` rather than `Z (a + 0)`, which can be achieved here by using the notation
+`x •[add_zero a] y`.
+
+Graded heterogeneous multiplications shall be useful in the development of
+cohomology theory:
+* If `K` and `L` are cochain complexes (indexed by `ℤ`) in a preadditive category,
+we may define `Cochain K L n` as families of morphisms `K.X p ⟶ L.X q` for all integers
+`p` and `q` such that `p + n = q`. If `M` is another cochain complex, we have a
+composition of cochains `HasGradedHMul (Cochain K L) (Cochain L M) (Cochain K M)`
+which is associative. This will play an important role in the construction
+of the triangulated structure on the homotopy category of cochain complexes of
+an additive category (TODO).
+* If `K` and `L` are objects of a triangulated category, "shifted homs" `K ⟶ L⟦n⟧`
+form a graded type indexed by `ℤ`, and the composition of these is also associative (TODO).
+* The composition of `Ext` groups in an abelian category can also be interpreted
+using an associative graded heterogeneous multiplication (TODO).
+
+-/
+
+universe u₁ u₂
+
+/-- A graded type indexed by `M` is a map `M → Type*` -/
+abbrev GradedType (M : Type u₁) := M → Type u₂
+
+variable {M G : Type u₁}
+
+/-- A graded heterogeneous multiplication from `X` and `Y` to `Z` (all are graded types
+indexed by `M`) consists of maps `X a → Y b → Z c` whenever `a + b = c` in `M`. -/
+class HasGradedHMul [Add M] (X Y : GradedType M) (Z : outParam (GradedType M)) where
+  /-- the heterogeneous multiplication -/
+  γhmul' (a b c : M) (h : a + b = c) (α : X a) (β : Y b) : Z c
+
+/-- When `[HasGradedHMul X Y Z]`, `x : X a`, `y : Y b` and `a + b = c`, this is the
+result in `Z c` of the graded heterogeneous multiplication of `x` and `y`. -/
+def HasGradedHMul.γhmul [Add M] {X Y : GradedType M} {Z : outParam (GradedType M)}
+    [HasGradedHMul X Y Z] {a b : M} (α : X a) (β : Y b) {c : M} (h : a + b = c) : Z c :=
+  γhmul' a b c h α β
+
+/-- When `[HasGradedHMul X Y Z]`, `x : X a`, `y : Y b` and `a + b = c`, this is the
+result in `Z c` of the graded heterogeneous multiplication of `x` and `y`. -/
+notation a " •[" h "] " b:80 => HasGradedHMul.γhmul a b h
+
+section
+
+variable [AddMonoid M] (X Y Z : GradedType M) (XY YZ XYZ : outParam (GradedType M))
+  [HasGradedHMul X Y XY] [HasGradedHMul Y Z YZ]
+  [HasGradedHMul X YZ XYZ] [HasGradedHMul XY Z XYZ]
+
+/-- The condition that graded heterogeneous multiplications are associative. -/
+class IsAssocGradedHMul : Prop where
+  /-- The associativity condition on graded heterogeneous multiplications. -/
+  γhmul_assoc : ∀ ⦃a b c : M⦄ (α : X a) (β : Y b) (γ : Z c) (ab bc abc : M)
+    (hab : a + b = ab) (hbc : b + c = bc) (habc : ab + c = abc),
+      (α •[hab] β) •[habc] γ =
+        α •[show a + bc = abc by rw [← hbc, ← add_assoc, hab, habc]] (β •[hbc] γ)
```

Comment:
This is essentially option 6 from https://link.springer.com/chapter/10.1007/978-3-031-16681-5_8#Sec4, which we ultimately rejected for the homogenous version.

Note that this typeclass is effectively [`GradedMonoid.GMulAction`](https://leanprover-community.github.io/mathlib4_docs//////Mathlib/Algebra/GradedMulAction.html#GradedMonoid.GMulAction), at least when `Y = Z`.